### ISSUE TITLE
Name check fails for XMPP server

### ISSUE THREAD

**Issue by joto** [drwetter/testssl.sh#1097]
Please make sure that you provide enough information so that we understand what your issue is about.

1. testssl version from the banner (testssl.sh -b 2>/dev/null | head -4 | tail -2)

    testssl       2.8rc3 from https://testssl.sh/dev/
    (33dcb43 2018-08-04 19:09:51 -- 1.562)

Using "OpenSSL 1.1.0f  25 May 2017"

(From Debian stable)

2. what exactly was happening, output is needed

Tested an XMPP server with `testssl --quiet --server_defaults -t xmpp --xmpphost topf.org jabber.topf.org:5222`

So the server itself has the name `jabber.topf.org`, but the jabber ids would be `user@topf.org`. In this case, the certificate should use `topf.org` as SN and it does.

testssl now says:
```
...
Trust (hostname)             certificate does not match supplied URI
...
```

Maybe using the `--xmpphost` option was the wrong thing to do in this case? The docs are not 100% clear here.

3. what did you expect instead?

I would have expected it to say that everything is okay.

4. steps to reproduce

see above

**Comment by dcooper16** [comment]
Hi @joto,

I can't access the server that you mentioned (probably a firewall issue), but it seems that you are correct. testssl.sh currently uses one set of matching rules for all types of servers - it checks for DNS names in the subjectAltName extension and it checks the common name attribute in the subject name.

According to Section 13.7.2.1 of [RFC 6120](https://tools.ietf.org/html/rfc6120#section-13.7.2.1), however, for XMPP servers there should also be a check against any XmppAddr names in the subjectAltName extension.

[RFC 6125](https://tools.ietf.org/html/rfc6125) also mentions checking for matching SRVName in the subjectAltName extension.

So, it seems that testssl.sh's rules for checking names in certificates are okay for HTTPS, and perhaps for some of the other non-HTTPS protocols it supports, but it isn't for XMPP and perhaps for some other protocols. To be complete, testssl.sh will need to add some protocol-specific matching rules.

I can try to help with this, but my ability may be limited, as I may be limited in what servers I can test against.

**Comment by drwetter** [comment]
This seems loosely related to #316. Best would be if we think about protocol specific hooks in ``certificate_info()``

OT:

> testssl 2.8rc3 from https://testssl.sh/dev/
> (33dcb43 2018-08-04 19:09:51 -- 1.562)

This is kind of confusing: 2.8 was released in 2017 and not 2018. Probably the commit 33dcb43 is related thereto. The old CVS commit number 1.562 (which I gave up after 2.8) means also it was forked 20 commits before release of 2.8.

Note: 2.8 is not maintained anymore by upstream=here. As David already indicated though, this issue exists in the current development version, too. 

Cheers, Dirk

PS: Thx, Jochen, for reporting. Long time no hear.
